extensions/oplogPopulator/OplogPopulator.js

```diff
@@ -77,30 +127,262 @@ class OplogPopulator {
+                }
+                break;
+            default:
+                this._logger.debug('Skipping unsupported change stream event', {
+                    method: 'OplogPopulator._handleChangeStreamChange',
+                    type: change.operationType,
+                    key: change.documentKey._id,
+                });
+                break;
+        }
+        this._logger.debug('Change stream event processed', {
+            method: 'OplogPopulator._handleChangeStreamChange',
+            key: change.documentKey._id,
+        });
+    }
+
+    /**
+     * Handler for change stream the error event
+     * it reestablishes the change stream when an error occurs
+     * @returns {undefined}
+     */
+    async _handleChangeStreamErrorEvent() {
+        this._logger.error('An error occured when listening to the change stream', {
+            method: 'OplogPopulator._setMetastoreChangeStream',
+        });
+        this._metastoreChangeStream.removeListener('change', this._handleChangeStreamChangeEvent.bind(this));
+        this._metastoreChangeStream.removeListener('error', this._handleChangeStreamErrorEvent.bind(this));
+        // closing and restarting the change stream
+        if (!this._metastoreChangeStream.isClosed()) {
+            await this._metastoreChangeStream.close();
+        }
+        this._setMetastoreChangeStream();
+    }
+
+    /**
+     * Initializes a change stream on the metastore collection
+     * @returns {undefined}
+     * @throws {InternalError}
+     */
+    _setMetastoreChangeStream() {
+        const changeStreamPipeline = [
+            {
+                $project: {
+                    '_id': 1,
+                    'operationType': 1,
+                    'documentKey._id': 1,
+                    'fullDocument.value': 1
+                },
+            },
+        ];
+        const changeStreamParams = { fullDocument: 'updateLookup' };
+        if (this._changeStreamResumeToken) {
+            /**
+             * using "startAfter" instead of "resumeAfter" to resume
+             * even after an invalid event
+             */
+            changeStreamParams.startAfter = this._changeStreamResumeToken;
+        }
+        try {
+            this._metastoreChangeStream = this._metastore.watch(changeStreamPipeline, changeStreamParams);
+            this._metastoreChangeStream.on('change', this._handleChangeStreamChangeEvent.bind(this));
+            this._metastoreChangeStream.on('error', this._handleChangeStreamErrorEvent.bind(this));
+            this._logger.debug('Change stream set', {
+                method: 'OplogPopulator._setMetastoreChangeStream',
+            });
+        } catch (err) {
+            this._logger.error('Error while setting the change stream', {
+                method: 'OplogPopulator._setMetastoreChangeStream',
+                error: err.message,
+            });
+            throw errors.InternalError.customizeDescription(err.message);
+        }
+    }
```

Comment:
looking at the code here, I think we may benefit from extracting a generic "change stream" wrapper class : e.g. to handle errors, resume token, ... (which is done here, and in a few other places)